### PR TITLE
feat: add row option to JustifiedLayout

### PR DESCRIPTION
**Pull request by daybrush**
## Issue

#382

### CODE REVIEW

File: src/layouts/JustifiedLayout.ts [hunk 1]
```diff
@@ -12,6 +25,7 @@ import { ILayout, IRectlProperties, SizeType, IInfiniteGridItem, IInfiniteGridGr
  * @param {Number} [options.minSize=0] Minimum size of item to be resized <ko> 아이템이 조정되는 최소 크기 </ko>
  * @param {Number} [options.maxSize=0] Maximum size of item to be resized <ko> 아이템이 조정되는 최대 크기 </ko>
  * @param {Array|Number} [options.column=[1, 8]] The number of items in a line <ko> 한 줄에 들어가는 아이템의 개수 </ko>
+ * @param {Array|Number} [options.line=0] The number or range of lines in a group, 0 is not set <ko>The number of lines in a group, 0 is not set</ko>
```
Comment (WoodNeck) [inline src/layouts/JustifiedLayout.ts:28]: 한글 설명이 반영되어 있지 않네요

File: src/utils.ts [hunk 2]
```diff
@@ -404,3 +404,10 @@ export function addClass(element: HTMLElement, className: string) {
 		element.className += ` ${className}`;
 	}
 }
+export function isObject(value: any): value is object {
+	return typeof value === "object";
+}
+
+export function getRangeCost(value: number, range: number[]) {
+	return Math.max(Math.max(value - range[1], 0), Math.max(range[0] - value, 0)) + 1;
```
Comment (WoodNeck) [inline src/utils.ts:412]: I think you can call just one `Math.max` in this case
```js
Math.max(value - range[1], range[0] - value, 0)
```

File: src/layouts/JustifiedLayout.ts [hunk 3]
```diff
@@ -124,39 +143,125 @@ class JustifiedLayout implements ILayout {
...
+		const [minColumn, maxColumn] = isObject(column) ? column : [column, column];
+		const [minLine, maxLine]: number[] = isObject(line) ? line : [line, line];
+		const lastNode = items.length;
 
-		return this._setStyle(items, path, outline, isAppend);
+		const getLinks = (prevLink: Link, prevNode: number) => {
+			const nextLinks: Link[] = [];
+			const { path, length: pathLength, cost } = prevLink;
+
+			if ((maxLine <= pathLength|| prevNode + minColumn > lastNode) && prevNode < lastNode) {
```
Comment (WoodNeck) [inline src/layouts/JustifiedLayout.ts:205]: There's no space before ||

**Comment by WoodNeck** [comment]
I might be misunderstanding something, but the opposite of `column` shouldn't be `row`? Why did you name it `line`?

**Comment by daybrush** [comment]
@WoodNeck 

You are right. I will change from line to row.

File: src/layouts/JustifiedLayout.ts [hunk 4]
```diff
@@ -124,39 +143,125 @@ class JustifiedLayout implements ILayout {
...
+		const size1Name = style.size1;
+		const size2Name = style.size2;
+		const column = this.options.column;
+		const row = this.options.row;
+		const [minColumn, maxColumn] = isObject(column) ? column : [column, column];
+		const [minRow, maxRow]: number[] = isObject(row) ? row : [row, row];
+		const lastNode = items.length;
 
-		return this._setStyle(items, path, outline, isAppend);
+		const getLinks = (prevLink: Link, prevNode: number) => {
```
Comment (mixed) [inline src/layouts/JustifiedLayout.ts:201]: How about use `extract method`?

File: src/layouts/JustifiedLayout.ts [hunk 5]
```diff
@@ -124,39 +143,125 @@ class JustifiedLayout implements ILayout {
...
+			const nextLinks: Link[] = [];
+			const { path, length: pathLength, cost } = prevLink;
+
+			if ((maxRow <= pathLength || prevNode + minColumn > lastNode) && prevNode < lastNode) {
```
Comment (mixed) [inline src/layouts/JustifiedLayout.ts:205]: What do these condition mean?

File: src/layouts/JustifiedLayout.ts [hunk 6]
```diff
@@ -124,39 +143,125 @@ class JustifiedLayout implements ILayout {
...
+					const nextCost = Math.abs(this._getCost(items, prevNode, nextNode, size1Name, size2Name));
+
+					nextLinks.push(...getLinks({
+						path: [...path, nextNode],
+						length: pathLength + 1,
+						cost: cost + nextCost,
+					}, nextNode));
+				}
+			}
+			nextLinks.sort((a, b) => {
```
Comment (mixed) [inline src/layouts/JustifiedLayout.ts:237]: I think that it is the logic that checks if the line has been overflow when placing it. How about creating a function with a name you give it meaning?

File: src/layouts/JustifiedLayout.ts [hunk 7]
```diff
@@ -124,39 +143,125 @@ class JustifiedLayout implements ILayout {
...
-				results[`${i}`] = Math.pow(cost, 2);
+				results[`${nextNode}`] = Math.pow(cost, 2);
 			}
 			return results;
 		};
 		// shortest path for items' total height.
-		const path = find_path(graph, `${startIndex}`, `${endIndex}`);
+		return find_path(graph, "0", `${lastNode}`);
+	}
+	private _getRowPath(items: IInfiniteGridItem[]) {
```
Comment (mixed) [inline src/layouts/JustifiedLayout.ts:191]: I think that This method seems to be large. How about breaking it down into small parts that work?

**Review by mixed — COMMENTED** [review]
🥇 🥇 🥇 🥇 🥇

File: src/layouts/JustifiedLayout.ts (hunk 4 above)
Comment (daybrush) [reply src/layouts/JustifiedLayout.ts:201]: There are many calls to the getLinks function. If separated, I thought there would be a lot of search for minColumn, maxColumn, minRow, and maxRow.

File: src/layouts/JustifiedLayout.ts (hunk 5 above)
Comment (daybrush) [reply src/layouts/JustifiedLayout.ts:205]: This is the case when the prevNode has not reached the last node, but the length of the path is greater than max row or less than minColumn to reach the last node.

File: src/layouts/JustifiedLayout.ts (hunk 6 above)
Comment (daybrush) [reply src/layouts/JustifiedLayout.ts:237]: right. If it does not overflow, the ranking is high, and if it overflows, the ranking is low.

Are you right to tell me to separate the sort function?

**Review by WoodNeck — APPROVED** [review]

File: src/layouts/JustifiedLayout.ts [hunk 8]
```diff
@@ -197,71 +208,86 @@ class JustifiedLayout implements ILayout {
...
+		const column = this.options.column;
+		const row = this.options.row;
+		const [minColumn, maxColumn] = isObject(column) ? column : [column, column];
+		const [minRow, maxRow]: number[] = isObject(row) ? row : [row, row];
+		const {
+			currentNode,
+			path,
+			length: pathLength,
+			cost
+		} = currentLink;
```
Comment (mixed) [inline src/layouts/JustifiedLayout.ts:255]: It seems to same code of `_getRowLink`. How about getting it as a parameter?

Comment (daybrush) [reply src/layouts/JustifiedLayout.ts:255]: So I put the getLink function inside the method to reduce duplicate code in the previous code. In the end, the parameters will be longer. Would it be okay?

**Review by mixed — APPROVED** [review]

File: src/layouts/JustifiedLayout.ts (hunk 8 above)
Comment (mixed) [reply src/layouts/JustifiedLayout.ts]: Longer parameters would be better than more points to manage. :)

**Pull request merged**